lib/files.c

```diff
@@ -568,3 +568,119 @@ bool files_load_bytes_from_file_or_stdin(const char *path, UINT16 *size, BYTE *b
+
+    UINT8 buffer[sizeof(*public)];
+    UINT16 size = sizeof(buffer);
+    bool res = files_load_bytes_from_path(path, buffer, &size);
+    if (!res) {
+        return false;
+    }
+
+    size_t offset = 0;
+    TSS2_RC rc = Tss2_MU_TPM2B_PUBLIC_Unmarshal(buffer, size, &offset, public);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing public structure: 0x%x", rc);
+        return false;
+    }
+
+    return rc == TPM2_RC_SUCCESS;
+}
+
+bool files_save_signature(TPMT_SIGNATURE *signature, const char *path) {
+
+    size_t offset = 0;
+    UINT8 buffer[sizeof(TPM2B_PUBLIC)];
+    TSS2_RC rc = Tss2_MU_TPMT_SIGNATURE_Marshal(signature, buffer, sizeof(buffer), &offset);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing signature structure: 0x%x", rc);
+        return false;
+    }
+
+    return files_save_bytes_to_file(path, buffer, offset);
+}
+
+bool files_load_signature(const char *path, TPMT_SIGNATURE *signature) {
+
+    UINT8 buffer[sizeof(*signature)];
+    UINT16 size = sizeof(buffer);
+    bool res = files_load_bytes_from_path(path, buffer, &size);
+    if (!res) {
+        return false;
+    }
+
+    size_t offset = 0;
+    TSS2_RC rc = Tss2_MU_TPMT_SIGNATURE_Unmarshal(buffer, size, &offset, signature);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing signature structure: 0x%x", rc);
+        return false;
+    }
+
+    return rc == TPM2_RC_SUCCESS;
+}
+
+bool files_save_ticket(TPMT_TK_VERIFIED *ticket, const char *path) {
+
+    size_t offset = 0;
+    UINT8 buffer[sizeof(TPM2B_PUBLIC)];
+    TSS2_RC rc = Tss2_MU_TPMT_TK_VERIFIED_Marshal(ticket, buffer, sizeof(buffer), &offset);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing signature structure: 0x%x", rc);
+        return false;
+    }
+
+    return files_save_bytes_to_file(path, buffer, offset);
+}
+
+bool files_load_ticket(const char *path, TPMT_TK_VERIFIED *ticket) {
+
+    UINT8 buffer[sizeof(*ticket)];
+    UINT16 size = sizeof(buffer);
+    bool res = files_load_bytes_from_path(path, buffer, &size);
+    if (!res) {
+        return false;
+    }
+
+    size_t offset = 0;
+    TSS2_RC rc = Tss2_MU_TPMT_TK_VERIFIED_Unmarshal(buffer, size, &offset, ticket);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing signature structure: 0x%x", rc);
+        return false;
+    }
+
+    return rc == TPM2_RC_SUCCESS;
+}
+
+
+bool files_load_sensitive(const char *path, TPM2B_SENSITIVE *sensitive) {
+
+    UINT8 buffer[sizeof(*sensitive)];
+    UINT16 size = sizeof(buffer);
+    bool res = files_load_bytes_from_path(path, buffer, &size);
+    if (!res) {
+        return false;
+    }
+
+    size_t offset = 0;
+    TSS2_RC rc = Tss2_MU_TPM2B_SENSITIVE_Unmarshal(buffer, size, &offset, sensitive);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error serializing signature structure: 0x%x", rc);
+        return false;
+    }
+
+    return rc == TPM2_RC_SUCCESS;
+}
```

Comment:
These could all be generated via macros....

Reply:
Indeed, all the files_{save,load}_* functions seems to be the same. The only thing that changes is the second parameter type and marshal/unmarshal function used (which are also generated via macros).